File: internal/hooks/hooks.go
package hooks

import (
	"tinyauth/internal/auth"
	"tinyauth/internal/providers"
	"tinyauth/internal/types"

	"github.com/gin-gonic/gin"
	"github.com/rs/zerolog/log"
)

type Hooks struct {
	Config    types.HooksConfig
	Auth      *auth.Auth
	Providers *providers.Providers
}

func NewHooks(config types.HooksConfig, auth *auth.Auth, providers *providers.Providers) *Hooks {
	return &Hooks{Config: config, Auth: auth, Providers: providers}
}

func (hooks *Hooks) UseUserContext(c *gin.Context) types.UserContext {
	cookie, err := hooks.Auth.GetSessionCookie(c)
	if err != nil || cookie.Provider == "" {
		log.Error().Err(err).Msg("Failed to get session cookie")
		return types.UserContext{}
	}

	provider := hooks.Providers.GetProvider(cookie.Provider)
	if provider == nil {
		return types.UserContext{}
	}

	return types.UserContext{
		Username:    cookie.Username,
		Name:        cookie.Name,
		Email:       cookie.Email,
		IsLoggedIn:  true,
		OAuth:       true,
		Provider:    cookie.Provider,
		OAuthGroups: cookie.OAuthGroups,
	}
}

File: internal/hooks/Hooks.cs
using System;
using Tinylinks.Auth;
using Tinylinks.Types;
using Tinylinks.Handlers; // using placeholder Providers

namespace Tinylinks.Hooks
{
    public class Hooks
    {
        public HooksConfig Config { get; set; }
        public Auth Auth { get; set; }
        public Providers Providers { get; set; }

        public Hooks(HooksConfig config, Auth auth, Providers providers)
        {
            Config = config;
            Auth = auth;
            Providers = providers;
        }

        public UserContext UseUserContext(Auth.Context ctx)
        {
            var cookie = Auth.GetSessionCookie(ctx);
            if (cookie == null || string.IsNullOrEmpty(cookie.Provider))
            {
                Console.Error.WriteLine("Session cookie missing or provider undefined");
                return new UserContext();
            }

            var provider = Providers.GetProvider(cookie.Provider);
            if (provider == null)
            {
                return new UserContext();
            }

            return new UserContext
            {
                Username = cookie.Username,
                Name = cookie.Name,
                Email = cookie.Email,
                IsLoggedIn = true,
                OAuth = true,
                Provider = cookie.Provider,
                OAuthGroups = cookie.OAuthGroups,
            };
        }
    }
}
